core/src/bls_sigverify/bls_sigverify_service.rs

```diff
@@ -0,0 +1,168 @@
+    Streamer(Box<StreamerError>),
+}
+
+impl<SendType> From<SendError<SendType>> for BLSSigVerifyServiceError<SendType> {
+    fn from(e: SendError<SendType>) -> Self {
+        Self::Send(Box::new(e))
+    }
+}
+
+impl<SendType> From<TrySendError<SendType>> for BLSSigVerifyServiceError<SendType> {
+    fn from(e: TrySendError<SendType>) -> Self {
+        Self::TrySend(Box::new(e))
+    }
+}
+
+impl<SendType> From<StreamerError> for BLSSigVerifyServiceError<SendType> {
+    fn from(e: StreamerError) -> Self {
+        Self::Streamer(Box::new(e))
+    }
+}
+
+type Result<T, SendType> = std::result::Result<T, BLSSigVerifyServiceError<SendType>>;
+
+pub struct BLSSigverifyService {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl BLSSigverifyService {
+    pub fn new(packet_receiver: Receiver<PacketBatch>, verifier: BLSSigVerifier) -> Self {
+        let thread_hdl = Self::verifier_service(packet_receiver, verifier);
+        Self { thread_hdl }
+    }
+
+    fn verifier<const K: usize>(
+        deduper: &Deduper<K, [u8]>,
+        recvr: &Receiver<PacketBatch>,
+        verifier: &mut BLSSigVerifier,
+        stats: &mut BLSPacketStats,
+    ) -> Result<(), ConsensusMessage> {
+        let (mut batches, num_packets, recv_duration) = streamer::recv_packet_batches(recvr)?;
+
+        let batches_len = batches.len();
+        debug!(
+            "@{:?} bls_verifier: verifying: {}",
+            timing::timestamp(),
+            num_packets,
+        );
+
+        let mut dedup_time = Measure::start("bls_sigverify_dedup_time");
+        let discard_or_dedup_fail =
+            deduper::dedup_packets_and_count_discards(deduper, &mut batches) as usize;
+        dedup_time.stop();
+
+        let mut verify_time = Measure::start("sigverify_batch_time");
+        verifier.verify_and_send_batches(batches)?;
+        verify_time.stop();
+
+        debug!(
+            "@{:?} verifier: done. batches: {} total verify time: {:?} verified: {} v/s {}",
+            timing::timestamp(),
+            batches_len,
+            verify_time.as_ms(),
+            num_packets,
+            (num_packets as f32 / verify_time.as_s())
+        );
+
+        stats
+            .recv_batches_us_hist
+            .increment(recv_duration.as_micros() as u64)
+            .unwrap();
+        stats
+            .verify_batches_pp_us_hist
+            .increment(verify_time.as_us() / (num_packets as u64))
+            .unwrap();
+        stats
+            .dedup_packets_pp_us_hist
+            .increment(dedup_time.as_us() / (num_packets as u64))
+            .unwrap();
+        stats.batches_hist.increment(batches_len as u64).unwrap();
+        stats.packets_hist.increment(num_packets as u64).unwrap();
+        stats.total_batches += batches_len;
+        stats.total_packets += num_packets;
+        stats.total_dedup += discard_or_dedup_fail;
+        stats.total_dedup_time_us += dedup_time.as_us() as usize;
+        stats.total_verify_time_us += verify_time.as_us() as usize;
```

Comment:
might be nice to try and hide this a bit in some sub-function. Seems like more than half the LOC in `verifier` are just stats stuff

Reply:
Sure. The final plan is to remove this file. But split these into a separate function for now.